Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,306 @@
+   * including the src_end.
+   */
+  template <typename SourceIterator>
+  void append(SourceIterator src_begin, SourceIterator src_end) {
+    auto copy_end = std::copy(src_begin, src_end, end());
+    size_ = static_cast<size_type>(copy_end - begin());
+  }
+
+  /**
+   * Appends a single value to the array.
+   */
+  void append(T&& value) {
+    *end() = std::forward<T>(value);
+    size_ += 1;
+  }
+
+  iterator begin() {
+    return contents_.begin();
+  }
+
+  const_iterator begin() const {
+    return contents_.begin();
+  }
+
+  iterator end() {
+    return begin() + size_;
+  }
+
+  const_iterator end() const {
+    return begin() + size_;
+  }
+
+ private:
+  array_type contents_;
+  size_type size_;
+
+  template <typename K, typename V, typename C>
+  friend class ArraySortedMap;
+};
+
+/**
+ * ArraySortedMap is a value type containing a map. It is immutable, but has
+ * methods to efficiently create new maps that are mutations of it.
+ */
+template <typename K, typename V, typename C = std::less<K>>
+class ArraySortedMap : public ArraySortedMapBase {
+ public:
+  typedef ArraySortedMap<K, V, C> this_type;
+  typedef KeyComparator<K, V, C> key_comparator_type;
+
+  /**
+   * The type of the entries stored in the map.
+   */
+  typedef std::pair<K, V> value_type;
+
+  /**
+   * The type of the fixed-size array containing entries of value_type.
+   */
+  typedef FixedArray<value_type, kFixedSize> array_type;
+  typedef typename array_type::const_iterator const_iterator;
+
+  typedef std::shared_ptr<const array_type> array_pointer;
+
+  /**
+   * Creates an empty ArraySortedMap.
+   */
+  explicit ArraySortedMap(const C& comparator = C())
+      : array_(EmptyArray()), key_comparator_(comparator) {
+  }
+
+  /**
+   * Creates an ArraySortedMap containing the given entries.
+   */
+  explicit ArraySortedMap(std::initializer_list<value_type> entries,
+                          const C& comparator = C())
+      : array_(), key_comparator_(comparator) {
+    assert(static_cast<size_type>(entries.size()) <= kFixedSize);
+
+    auto array = std::make_shared<array_type>();
+    array->append(entries.begin(), entries.end());
+    array_ = array;
+  }
+
+  /**
+   * Creates a new map identical to this one, but with a key-value pair added or
+   * updated.
+   *
+   * @param key The key to insert/update.
+   * @param value The value to associate with the key.
+   * @return A new dictionary with the added/updated value.
+   */
+  this_type insert(const K& key, const V& value) const {
+    value_type pair(key, value);
+
+    const_iterator current_end = end();
+    const_iterator pos = LowerBound(pair.first);
+    bool replacing_entry = false;
+
+    if (pos != current_end) {
+      // LowerBound found an entry where pos->first >= pair.first. Reversing the
+      // argument order here tests pair.first < pos->first.
+      bool keys_equal = !key_comparator_(pair.first, *pos);
```

Comment:
Optional nit: consider replacing keys_equal with replacing_entry. I think it will make this fragment a little clearer due to reduced length, but it's a matter of taste:

```
bool replacing_entry = !key_comparator_(key, *pos);
if (replacing_entry && value == pos->second) {
  return *this;
}
```

Reply:
I'm glad you like it!

In general, I tend to have a lot of minor comments, so I always try to make it clear that they're optional, unless there's a bug/some other issue involved.